src/cmt_decode_msgpack.c

```diff
@@ -0,0 +1,647 @@
+    }
+
+    return CMT_DECODE_MSGPACK_SUCCESS;
+}
+
+static int unpack_metric_label(mpack_reader_t *reader, size_t index, void *context)
+{
+    struct cmt_msgpack_decode_context *decode_context;
+
+    if (NULL == reader  || 
+        NULL == context ) {
+        return CMT_DECODE_MSGPACK_INVALID_ARGUMENT_ERROR;
+    }
+
+    decode_context = (struct cmt_msgpack_decode_context *) context;
+
+    return unpack_label(reader, 
+                        index, 
+                        &decode_context->unique_label_list, 
+                        &decode_context->metric->labels);
+}
+
+static int unpack_metric_ts(mpack_reader_t *reader, size_t index, void *context)
+{
+    struct cmt_msgpack_decode_context *decode_context;
+
+    if (NULL == reader  || 
+        NULL == context ) {
+        return CMT_DECODE_MSGPACK_INVALID_ARGUMENT_ERROR;
+    }
+
+    decode_context = (struct cmt_msgpack_decode_context *) context;
+
+    return cmt_mpack_consume_uint_tag(reader, &decode_context->metric->timestamp);
+}
+
+static int unpack_metric_value(mpack_reader_t *reader, size_t index, void *context)
+{
+    double                             value;
+    int                                result;    
+    struct cmt_msgpack_decode_context *decode_context;
+
+    if (NULL == reader  || 
+        NULL == context ) {
+        return CMT_DECODE_MSGPACK_INVALID_ARGUMENT_ERROR;
+    }
+
+    decode_context = (struct cmt_msgpack_decode_context *) context;
+
+    result = cmt_mpack_consume_double_tag(reader, &value);
+
+    if(CMT_DECODE_MSGPACK_SUCCESS == result) {
+        decode_context->metric->val = cmt_math_d64_to_uint64(value);
+    }
+
+    return result;
+}
+
+static int unpack_metric_labels(mpack_reader_t *reader, size_t index, void *context)
+{
+    if (NULL == reader || 
+        NULL == context) {
+        return CMT_DECODE_MSGPACK_INVALID_ARGUMENT_ERROR;
+    }
+
+    return cmt_mpack_unpack_array(reader, 
+                                  unpack_metric_label, 
+                                  context);
+}
+
+static int unpack_metric(mpack_reader_t *reader, 
+                         struct cmt_msgpack_decode_context *decode_context, 
+                         struct cmt_metric **out_metric)
+{
+    int                                   result;
+    struct cmt_metric                    *metric;
+    struct cmt_mpack_map_entry_callback_t callbacks[] = \
+        {
+            {"ts",     unpack_metric_ts},
+            {"value",  unpack_metric_value},
+            {"labels", unpack_metric_labels},
+            {NULL,     NULL}
+        };
+
+    if (NULL == reader         || 
+        NULL == decode_context ||
+        NULL == out_metric) {
+        return CMT_DECODE_MSGPACK_INVALID_ARGUMENT_ERROR;
+    }
+
+    /* Maybe we could move this cmt_metric constructor code to its own file, add a 
+     * destructor and update map_metric_create and map_metric_destroy to use them right?
+     */
+
+    metric = malloc(sizeof(struct cmt_metric));
```

Comment:
use calloc(), so you can get rid of the next memset()

Reply:
I'll make that change now, it makes sense.